File: yang/you/wallpaper-miniapp/scripts/generate-icons.js
/**
 * 生成小程序 Tab 图标 - 纯 Node.js 无依赖版本
 * 使用内置的最小有效 PNG 格式
 */
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

const DIR = path.join(__dirname, '..', 'images', 'tab');
if (!fs.existsSync(DIR)) fs.mkdirSync(DIR, { recursive: true });

const SIZE = 81;

// 创建一个 RGBA 像素数组并编码为 PNG
function createIconPNG(drawFn) {
  const pixels = Buffer.alloc(SIZE * SIZE * 4, 0);
  drawFn(pixels, SIZE);
  
  // PNG 文件结构
  const signature = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);
  
  // IHDR chunk
  const ihdr_data = Buffer.alloc(13);
  ihdr_data.writeUInt32BE(SIZE, 0);  // width
  ihdr_data.writeUInt32BE(SIZE, 4);  // height
  ihdr_data[8] = 8;   // bit depth
  ihdr_data[9] = 6;   // color type (RGBA)
  ihdr_data[10] = 0;  // compression
  ihdr_data[11] = 0;  // filter
  ihdr_data[12] = 0;  // interlace
  
  // IDAT: 每行前加 filter byte (0 = None)
  const rawData = Buffer.alloc(SIZE * (1 + SIZE * 4));
  for (let y = 0; y < SIZE; y++) {
    rawData[y * (1 + SIZE * 4)] = 0; // filter byte
    pixels.copy(rawData, y * (1 + SIZE * 4) + 1, y * SIZE * 4, (y + 1) * SIZE * 4);
  }
  const idat_data = zlib.deflateSync(rawData);
  
  function chunk(type, data) {
    const len = Buffer.alloc(4);
    len.writeUInt32BE(data.length, 0);
    const typeB = Buffer.from(type, 'ascii');
    const crcData = Buffer.concat([typeB, data]);
    const crc = Buffer.alloc(4);
    crc.writeUInt32BE(crc32(crcData), 0);
    return Buffer.concat([len, typeB, data, crc]);
  }
  
  return Buffer.concat([
    signature,
    chunk('IHDR', ihdr_data),
    chunk('IDAT', idat_data),
    chunk('IEND', Buffer.alloc(0))
  ]);
}

// CRC32 实现
function crc32(buf) {
  let table = [], crc = 0xFFFFFFFF;
  for (let i = 0; i < 256; i++) { let c = i; for (let j = 0; j < 8; j++) c = c & 1 ? 0xEDB88320 ^ (c >>> 1) : c >>> 1; table[i] = c; }
  for (let i = 0; i < buf.length; i++) crc = table[(crc ^ buf[i]) & 0xFF] ^ (crc >>> 8);
  return (crc ^ 0xFFFFFFFF) >>> 0;
}

// 设置像素 (x, y) 为 rgba 颜色
function setPixel(px, w, x, y, r, g, b, a = 255) {
  if (x < 0 || x >= w || y < 0 || y >= w) return;
  const idx = (y * w + x) * 4;
  px[idx] = r; px[idx + 1] = g; px[idx + 2] = b; px[idx + 3] = a;
}

// 画填充圆
function fillCircle(px, w, cx, cy, radius, r, g, b, a = 255) {
  for (let y = Math.max(0, Math.floor(cy - radius)); y <= Math.min(w - 1, Math.ceil(cy + radius)); y++) {
    for (let x = Math.max(0, Math.floor(cx - radius)); x <= Math.min(w - 1, Math.ceil(cx + radius)); x++) {
      if ((x - cx) ** 2 + (y - cy) ** 2 <= radius ** 2) setPixel(px, w, x, y, r, g, b, a);
    }
  }
}

// 画线段 (Bresenham)
function drawLine(px, w, x0, y0, x1, y1, r, g, b, thickness = 6) {
  const dx = Math.abs(x1 - x0), dy = Math.abs(y1 - y0);
  const sx = x0 < x1 ? 1 : -1, sy = y0 < y1 ? 1 : -1;
  let err = dx - dy;
  while (true) {
    fillCircle(px, w, x0, y0, thickness / 2, r, g, b);
    if (x0 === x1 && y0 === y1) break;
    const e2 = 2 * err;
    if (e2 > -dy) { err -= dy; x0 += sx; }
    if (e2 < dx) { err += dx; y0 += sy; }
  }
}

// 画填充矩形
function fillRect(px, w, x, y, rw, rh, r, g, b, a = 255) {
  for (let py = Math.max(0, y); py < Math.min(w, y + rh); py++) {
    for (let px2 = Math.max(0, x); px2 < Math.min(w, x + rw); px2++) setPixel(px, w, px2, py, r, g, b, a);
  }
}

// 画填充三角形
function fillTriangle(px, w, x1, y1, x2, y2, x3, y3, r, g, b) {
  const minY = Math.max(0, Math.floor(Math.min(y1, y2, y3)));
  const maxY = Math.min(w - 1, Math.ceil(Math.max(y1, y2, y3)));
  for (let y = minY; y <= maxY; y++) {
    const intersections = [];
    [[x1, y1], [x2, y2], [x3, y3]].forEach(([px0, py0], i, arr) => {
      const [pxn, pyn] = arr[(i + 1) % 3];
      if ((py0 <= y && pyn > y) || (pyn <= y && py0 > y)) {
        const t = (y - py0) / (pyn - py0);
        intersections.push(Math.round(px0 + t * (pxn - px0)));
      }
    });
    if (intersections.length === 2) {
      const [xs, xe] = [Math.min(...intersections), Math.max(...intersections)];
      for (let x = Math.max(0, xs); x <= Math.min(w - 1, xe); x++) setPixel(px, w, x, y, r, g, b);
    }
  }
}

const C = Math.floor(SIZE / 2);

// ====== 图标绘制函数 ======

function drawHome(active) {
  const color = active ? [7, 193, 96] : [153, 153, 153]; // #07C160 or #999999
  return (px, w) => {
    // 房子主体三角形屋顶
    fillTriangle(px, w, C, 14, C - 28, 40, C + 28, 40, ...color);
    // 房子矩形主体
    fillRect(px, w, C - 28, 38, 56, 28, ...color);
    // 门（白色/透明）
    fillRect(px, w, C - 9, 48, 18, 18, 255, 255, 255);
    // 窗户左
    fillRect(px, w, C - 24, 44, 10, 10, 255, 255, 255);
    // 窗户右
    fillRect(px, w, C + 14, 44, 10, 10, 255, 255, 255);
  };
}

function drawCategory(active) {
  const color = active ? [7, 193, 96] : [153, 153, 153];
  return (px, w) => {
    const s = 17, gap = 5;
    // 四个方块
    fillRect(px, w, C - s - gap, C - s - gap, s, s, ...color);
    fillRect(px, w, C + gap, C - s - gap, s, s, ...color);
    fillRect(px, w, C - s - gap, C + gap, s, s, ...color);
    fillRect(px, w, C + gap, C + gap, s, s, ...color);
  };
}

function drawSearch(active) {
  const color = active ? [7, 193, 96] : [153, 153, 153];
  return (px, w) => {
    // 圆形镜片
    fillCircle(px, w, C - 4, C - 4, 20, ...color);
    // 内部镂空（白色）
    fillCircle(px, w, C - 4, C - 4, 13, 255, 255, 255);
    // 手柄
    const angle = Math.PI / 4;
    const handleStartX = C - 4 + Math.cos(angle) * 18;
    const handleStartY = C - 4 + Math.sin(angle) * 18;
    drawLine(px, w, Math.round(handleStartX), Math.round(handleStartY), C + 16, C + 16, ...color, 7);
  };
}

function drawMine(active) {
  const color = active ? [7, 193, 96] : [153, 153, 153];
  return (px, w) => {
    // 头
    fillCircle(px, w, C, 25, 15, ...color);
    // 身体半圆
    fillCircle(px, w, C, 58, 24, ...color);
    // 肩部以上白色区域（形成半圆形身体效果）
    for (let y = 0; y < 42; y++) {
      for (let x = 0; x < w; x++) {
        const dx = x - C, dy = y - 58;
        if (dx * dx + dy * dy < 24 * 24 && y < 42) {
          setPixel(px, w, x, y, 255, 255, 255, 0);
        }
      }
    }
  };
}

// ====== 生成所有图标 ======
console.log('🎨 正在生成 Tab 图标...\n');

const icons = [
  ['home', drawHome(false)],
  ['home-active', drawHome(true)],
  ['category', drawCategory(false)],
  ['category-active', drawCategory(true)],
  ['search', drawSearch(false)],
  ['search-active', drawSearch(true)],
  ['mine', drawMine(false)],
  ['mine-active', drawMine(true)]
];

icons.forEach(([name, drawFn]) => {
  const png = createIconPNG(drawFn);
  fs.writeFileSync(path.join(DIR, `${name}.png`), png);
  console.log(`  ✅ ${name}.png (${png.length} bytes)`);
});

console.log('\n🎉 全部完成！共生成 8 个 Tab 图标');

File: yang/you/wallpaper-miniapp/scripts/generate_icons.py
"""
生成小程序 Tab 图标 - Python 无依赖版本
运行: python3 scripts/generate-icons.py
"""
import struct, zlib, os, math

SIZE = 81
DIR = os.path.join(os.path.dirname(__file__), '..', 'images', 'tab')
os.makedirs(DIR, exist_ok=True)

def create_png(pixels):
    """从 RGBA 像素数组 (bytes) 生成 PNG bytes"""
    def chunk(ctype, data):
        c = ctype.encode('ascii') if isinstance(ctype, str) else ctype
        body = c + data
        crc = zlib.crc32(body) & 0xffffffff
        return struct.pack('>I', len(data)) + body + struct.pack('>I', crc)
    
    sig = b'\x89PNG\r\n\x1a\n'
    ihdr = struct.pack('>IIBBBBB', SIZE, SIZE, 8, 6, 0, 0, 0)
    raw = b''
    for y in range(SIZE):
        raw += b'\x00' + pixels[y * SIZE * 4:(y + 1) * SIZE * 4]
    
    return sig + chunk(b'IHDR', ihdr) + chunk(b'IDAT', zlib.compress(raw)) + chunk(b'IEND', b'')

class PixelBuffer:
    def __init__(self, size=SIZE):
        self.size = size
        self.px = bytearray(size * size * 4)
    
    def set(self, x, y, r, g, b, a=255):
        if 0 <= x < self.size and 0 <= y < self.size:
            idx = (y * self.size + x) * 4
            self.px[idx:idx+4] = [r, g, b, a]
    
    def fill_rect(self, x, y, w, h, r, g, b, a=255):
        for py in range(max(0,y), min(self.size, y+h)):
            for px in range(max(0,x), min(self.size, x+w)):
                self.set(px, py, r, g, b, a)
    
    def fill_circle(self, cx, cy, rad, r, g, b, a=255):
        for py in range(max(0,int(cy-rad)), min(self.size, int(cy+rad)+1)):
            for px in range(max(0,int(cx-rad)), min(self.size, int(cx-rad)+1)):
                if (px-cx)**2 + (py-cy)**2 <= rad*rad:
                    self.set(px, py, r, g, b, a)
    
    def fill_triangle(self, x1, y1, x2, y2, x3, y3, r, g, b):
        pts = [(x1,y1),(x2,y2),(x3,y3)]
        ymin = max(0, int(min(y for _,y in pts)))
        ymax = min(self.size-1, int(max(y for _,y in pts)))
        for y in range(ymin, ymax+1):
            inters = []
            for i in range(3):
                (ax, ay), (bx, by) = pts[i], pts[(i+1)%3]
                if (ay <= y < by) or (by <= y < ay):
                    t = (y - ay) / (by - ay)
                    inters.append(int(ax + t*(bx - ax)))
            if len(inters) == 2:
                for x in range(max(0, min(inters)), min(self.size, max(inters))+1):
                    self.set(x, y, r, g, b)
    
    def draw_line(self, x0, y0, x1, y1, r, g, b, thick=6):
        dx, dy = abs(x1-x0), abs(y1-y0)
        sx, sy = (1 if x0<x1 else -1), (1 if y0<y1 else -1)
        err = dx - dy
        while True:
            self.fill_circle(x0, y0, thick/2, r, g, b)
            if x0 == x1 and y0 == y1: break
            e2 = 2*err
            if e2 > -dy: err -= dy; x0 += sx
            if e2 < dx: err += dx; y0 += sy
    
    def to_bytes(self): return bytes(self.px)

C = SIZE // 2

def draw_home(active):
    color = (7, 193, 96) if active else (153, 153, 153)
    buf = PixelBuffer()
    buf.fill_triangle(C, 14, C-28, 40, C+28, 40, *color)
    buf.fill_rect(C-28, 38, 56, 28, *color)
    # 门
    buf.fill_rect(C-9, 48, 18, 18, 255, 255, 255)
    # 窗户
    buf.fill_rect(C-24, 44, 10, 10, 255, 255, 255)
    buf.fill_rect(C+14, 44, 10, 10, 255, 255, 255)
    return create_png(buf.to_bytes())

def draw_category(active):
    color = (7, 193, 96) if active else (153, 153, 153)
    buf = PixelBuffer()
    s, gap = 17, 5
    buf.fill_rect(C-s-gap, C-s-gap, s, s, *color)
    buf.fill_rect(C+gap, C-s-gap, s, s, *color)
    buf.fill_rect(C-s-gap, C+gap, s, s, *color)
    buf.fill_rect(C+gap, C+gap, s, s, *color)
    return create_png(buf.to_bytes())

def draw_search(active):
    color = (7, 193, 96) if active else (153, 153, 153)
    buf = PixelBuffer()
    # 镜片圆环
    buf.fill_circle(C-4, C-4, 20, *color)
    buf.fill_circle(C-4, C-4, 12, 255, 255, 255)
    # 手柄
    import math
    angle = math.pi / 4
    hsx = C-4 + round(math.cos(angle)*18)
    hsy = C-4 + round(math.sin(angle)*18)
    buf.draw_line(hsx, hsy, C+16, C+16, *color, 7)
    return create_png(buf.to_bytes())

def draw_mine(active):
    color = (7, 193, 96) if active else (153, 153, 153)
    buf = PixelBuffer()
    # 头
    buf.fill_circle(C, 25, 15, *color)
    # 身体（半圆效果）
    buf.fill_circle(C, 58, 24, *color)
    # 清除身体上半部分形成半圆
    for y in range(42):
        for x in range(SIZE):
            if (x-C)**2 + (y-58)**2 < 24**2:
                buf.set(x, y, 255, 255, 255, 0)
    return create_png(buf.to_bytes())

icons = [
    ('home', draw_home(False)),
    ('home-active', draw_home(True)),
    ('category', draw_category(False)),
    ('category-active', draw_category(True)),
    ('search', draw_search(False)),
    ('search-active', draw_search(True)),
    ('mine', draw_mine(False)),
    ('mine-active', draw_mine(True)),
]

print('🎨 正在生成 Tab 图标...\n')
for name, data in icons:
    path = os.path.join(DIR, f'{name}.png')
    with open(path, 'wb') as f:
        f.write(data)
    print(f'  ✅ {name}.png ({len(data)} bytes)')

print('\n🎉 全部完成！共生成 8 个 Tab 图标')
